### PR TITLE
Removing subscriber concurency

### PR DESCRIPTION
**Pull request by aabouman**

### CODE REVIEW

File: .github/workflows/CI.yml [hunk 1]
```diff
@@ -2,7 +2,7 @@ name: CI
 on:
   push:
     branches:
-      - main
+      - master
```
Comment (bjack205) [inline .github/workflows/CI.yml]: Why did you change this? The default branch of this repo is `main`, not `master`

File: deps/src/get_serial.cmake [hunk 2]
```diff
@@ -0,0 +1,48 @@
+cmake_minimum_required(VERSION 3.0.0)
```
Comment (bjack205) [inline deps/src/get_serial.cmake]: you only need this at the top of the root file

File: deps/src/get_serial.cmake [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+########################################
+# Download libserialports dependencies
+########################################
+include(ExternalProject)  # pull in the `ExternalProject` CMake module
+include(CMakePrintHelpers)
+
+# get_filename_component(SERIALPORT_PRE ${CMAKE_CURRENT_BINARY_DIR}/_deps/libserialport ABSOLUTE)
```
Comment (bjack205) [inline deps/src/get_serial.cmake]: nit: clean out these commented lines

File: deps/src/get_serial.cmake [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+# Set location of downloaded serialport libary
+if (APPLE)
+    set(SERIALPORT_LIB "${SERIALPORT_PREFIX}/src/serialport-build/.libs/libserialport.dylib")
+elseif (UNIX)
+    set(SERIALPORT_LIB "${SERIALPORT_PREFIX}/src/serialport-build/.libs/libserialport.so")
+endif()
+
+# Make library and make sure that serialport is built before including directories
+# and seting properites
```
Comment (bjack205) [inline deps/src/get_serial.cmake]: nit: typo in "setting"

File: deps/src/get_zmq.cmake [hunk 5]
```diff
@@ -0,0 +1,29 @@
+cmake_minimum_required(VERSION 3.0.0)
```
Comment (bjack205) [inline deps/src/get_zmq.cmake]: ditto previous

File: deps/src/get_zmq.cmake [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    FetchContent_Populate(libzmq)
+    set(ZMQ_BUILD_TESTS OFF CACHE BOOL "Test suite for libzmq")
+    # To enable building on macOS, see github issue https://github.com/zeromq/libzmq/issues/4085
+    set(WITH_TLS OFF CACHE INTERNAL "Disable TLS support")
+    add_subdirectory(${libzmq_SOURCE_DIR} ${libzmq_BINARY_DIR})
+endif()
+
+# Modify the `libzmq` target generated by it's build system to include the zmq header
+# files as part of its public interface
+target_include_directories(libzmq
```
Comment (bjack205) [inline deps/src/get_zmq.cmake:23]: Since you add the include directories to the target here, you shouldn't need to do it when linking against the library above. CMake will automatically add the include directories from any public or interface include directories of linked targets.

File: deps/src/relay_launch.c [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+#include <stdbool.h>
+
+#include "serial_relay.h"
+
+
+int main(int argc, char **argv)
+{
+    if (argc != 5)
+    {
+        fprintf(stderr, "Expects 4 additional command line arguments: (port_name, baudrate, msg_size, sub_endpoint, pub_endpoint)", errno);
```
Comment (bjack205) [inline deps/src/relay_launch.c]: You have 5 listed?

Comment (bjack205) [reply deps/src/relay_launch.c]: Looks like you need to remove `msg_size` from this list?

File: deps/src/relay_launch.c [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    const char *port_name;
+    int baudrate;
+    // size_t msg_size;
+    const char *sub_endpoint;
+    const char *pub_endpoint;
+
+    port_name = argv[1];
+    if (sscanf(argv[2], "%i", &baudrate) != 1)
+    {
+        fprintf(stderr, "error - not an integer");
```
Comment (bjack205) [inline deps/src/relay_launch.c]: You should provide a more helpful error message (saying that "baudrate" isn't an integer). And if it failed to parse the baudrate and you don't have a default, you should terminate the program instead of continuing.

File: deps/src/relay_launch.c [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    sub_endpoint = argv[3];
+    pub_endpoint = argv[4];
+
+    relay_launch(port_name,
+                 baudrate,
+                //  msg_size,
+                 sub_endpoint,
+                 pub_endpoint);
+
+    return 1;
```
Comment (bjack205) [inline deps/src/relay_launch.c]: You have your returns mixed up. `main` should return 0 upon success and 1 if there's an error.

File: deps/src/serial_relay.c [hunk 10]
```diff
@@ -0,0 +1,361 @@
+#include <string.h>
+#include "serial_relay.h"
+
+#define PORT_BUFFER_SIZE 1024
```
Comment (bjack205) [inline deps/src/serial_relay.c:4]: you could make this an option in CMake so the user can specify it at build time

File: deps/src/serial_relay.c [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+        size_t msg_size = zmq_msg_size(&msg);
+        if (msg_size > PORT_BUFFER_SIZE)
+        {
+            fprintf(stderr, "Message size is too large to write to serial port!");
+            return SR_ERR_MEM;
+        }
+
+        // Copy the msg into the buffer and free ZMQ message
+        memcpy(relay->msg_sub_buffer, zmq_msg_data(&msg), msg_size);
+        flag = check_zmq(zmq_msg_close(&msg));
```
Comment (bjack205) [inline deps/src/serial_relay.c:238]: Should this be outside the `if` statement?

File: deps/src/serial_relay.c [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+
+    flag = check_zmq(zmq_ctx_destroy(relay->context));
+    if (flag != SR_OK) return flag;
+
+    // Close out the serial port
+    flag = check_serial(sp_close(relay->port));
+    if (flag != SR_OK) return flag;
+
+    sp_free_port(relay->port);
+
```
Comment (bjack205) [inline deps/src/serial_relay.c:288]: no `free` on the relay itself? It's usually  best practice to mirror the allocs in the initializer.

File: deps/src/serial_relay.c [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+    if (relay == NULL)
+    {
+        fprintf(stderr, "Failed to initialize serial-zmq relay!");
+        return;
+    }
+    else
+    {
+        while (true)
+        {
+            // Add checks here to make sure no errors were thrown
```
Comment (bjack205) [inline deps/src/serial_relay.c]: is this a TODO? If so, add TODO in front of the comment and ideally create a ticket for it so you come back to it

File: deps/src/serial_relay.c [hunk 14]
```diff
@@ -0,0 +1,361 @@
...
+            relay_read(relay);
+            relay_write(relay);
+        }
+        close_relay(relay);
```
Comment (bjack205) [inline deps/src/serial_relay.c:321]: memory leak? You never free the relay. Add this to the `close_relay` method.

File: deps/src/serial_relay.c [hunk 15]
```diff
@@ -0,0 +1,361 @@
...
+        return;
+    }
+}
+
+
+// Add a check valid serial_relay type ie not all nulls port is open etc.
```
Comment (bjack205) [inline deps/src/serial_relay.c]: TODO?

File: deps/src/serial_relay.c [hunk 16]
```diff
@@ -0,0 +1,361 @@
...
+enum sr_return check_zmq(int rc)
+{
+    if (rc == -1)
+    {
+        fprintf(stderr, "Error occurred during zmq_init(): %s\n", zmq_strerror(zmq_errno()));
```
Comment (bjack205) [inline deps/src/serial_relay.c]: Fix this error message? This is used to check more than `zmq_init`

File: deps/src/serial_relay.h [hunk 17]
```diff
@@ -0,0 +1,37 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+
+#include <libserialport.h>
+#include <zmq.h>
+
+#define DEBUG 1
```
Comment (bjack205) [inline deps/src/serial_relay.h]: It'd be better to make this a CMake variable so the user can more easily control it from the build system.

File: deps/src/serial_relay.h [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+    SR_ERR_SP = -1,
+    // libzmq error
+    SR_ERR_ZMQ = -2,
+    // A memory allocation failure
+    SR_ERR_MEM = -3,
+};
+
+// Open a serial port in read write mode and the associated ZMQ ports.
+// If any issue arises closes everything and returns a NULL pointer.
+void *open_relay(const char *port_name,
```
Comment (bjack205) [inline deps/src/serial_relay.h:35]: Add Doxygen-style comments to the interface

**Review by bjack205 — COMMENTED** [review]
Some general comments on the serial relay code. Overall it looked really good! Good C hygiene. Mostly just a few nitpicks here and there, other than the memory leak. 

If you haven't run this through valgrind, I'd highly recommend it.

File: src/node.jl [hunk 19]
```diff
@@ -199,10 +191,9 @@ getoptions(node::Node) = getIO(node).opts
 getflags(node::Node) = getIO(node).flags
 getrate(node::Node)::Float64 = getoptions(node).rate
 function isnodedone(node::Node)::Bool
-    nodeio = getIO(node)
-    all_running = all(isrunning.(nodeio.subs))
-    return getflags(node).should_finish[] || !all_running
+    return getflags(node).should_finish[]
```
Comment (bjack205) [inline src/node.jl:194]: Not quite true. Maybe have a flag that gets set to true when the node exits? For example, what if it somehow hangs between `should_finish` getting set to true and it actually finishing?

File: src/node.jl [hunk 20]
```diff
@@ -279,17 +270,18 @@ function launch(node::Node)
...
         startup(node)
 
         getflags(node).is_running[] = true
 
         @rate while !isnodedone(node)
+            # Check the subscribers for new messages
             compute(node)
 
+            # Check the subscribers for new messages
+            poll_subscribers(node)
```
Comment (bjack205) [inline src/node.jl]: Is there a reason `poll_subscribers` comes after `compute`? Won't that mean the first time `compute` is called it won't have any data from the subscribers?

File: src/serial_relay.jl [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+    for port_name in keys(SERIAL_PORTS)
+        @info "Closing down Serial-ZMQ relay running on serialport: $port_name"
+
+        close(SERIAL_PORTS[port_name])
+        delete!(SERIAL_PORTS, port_name)
+    end
+end
+
+# Make sure all processes are cleaned up
+Base.atexit(closeall_relays)
```
Comment (bjack205) [inline src/serial_relay.jl:64]: nice

File: src/utils.jl [hunk 22]
```diff
@@ -1,5 +1,7 @@
 ZMQ_CONFLATE = 54
 
+const MercuryMessage = Union{ProtoBuf.ProtoType,AbstractVector{UInt8}}
```
Comment (bjack205) [inline src/utils.jl]: This should go in `Mercury.jl` instead or being buried in utils.

**Review by bjack205 — CHANGES_REQUESTED** [review]
Overall looks really good. Getting rid of the serial stuff really cleans it up. After making the few change go ahead and merge it.

File: .github/workflows/CI.yml (hunk 1 above)
Comment (aabouman) [reply .github/workflows/CI.yml]: Oh my bad I copied that from the serial relay repo where I was testing ill change that back to main.

File: deps/src/serial_relay.c (hunk 11 above)
Comment (aabouman) [reply deps/src/serial_relay.c:238]: I don't think so, we only want to construct the message and copy bytes if the we have recieved a message and if the message is small enough to write to the serial port.
